### PR TITLE
`Programming exercises`: Indicate who reset repository

### LINKED ISSUE #11807: Indicate who used the "Reset Repository" feature
### Is your feature request related to a problem?

Artemis allows students to reset their repository in programming exercises. For individual exercises, this works fine.
However, for team exercises any of the team members can do this and there is currently no indication whatsoever who did this. The corresponding commit is attributed to Artemis and there is not even an entry in the logs about this.
This means that a malicious team member could reset the exercise just before the deadline and there is no real way to figure out who it was. This almost happened during a real team exercise.

### Describe the solution you'd like

The minimum here would be to at least add some log output to the `ProgrammingExerciseParticipationResource#resetRepository` endpoint.
Even better would be some way to communicate who did this to the other team members and instructors. For this there are several viable approaches. The simplest one in my opinion would be to just attribute the commit to the team member in some way (Either as primary or as co-author).

### Describe alternatives you've considered

- Just remove the reset repository feature for team exercises, but this is not preferred, as it can still be useful
- Require all team members to agree to a repository reset, but this is too complex to integrate
- Display who performed this action in some new web UI, but this would also be too complex for just this feature

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by JTNing**
### Summary 
This PR addresses the issue in group programming exercises where it was previously not possible to track who reset the repository.

#### General

- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server.
- [x] This is a small issue that I tested locally and was confirmed by another developer on a test server.
- [x] Language: I followed the [guidelines for inclusive, diversity-sensitive, and appreciative language](https://docs.artemis.tum.de/developer/guidelines/language).
- [x] I chose a title conforming to the [naming conventions for pull requests](https://docs.artemis.tum.de/developer/development-process#pr-naming-conventions).

#### Server
- [x] **Important**: I implemented the changes with a [very good performance](https://docs.artemis.tum.de/developer/guidelines/performance) and prevented too many (unnecessary) and too complex database calls.
- [x] I **strictly** followed the principle of **data economy** for all database calls.
- [x] I **strictly** followed the [server coding and design guidelines](https://docs.artemis.tum.de/developer/guidelines/server-development) and the [REST API guidelines](https://docs.artemis.tum.de/developer/guidelines/rest-api).
- [x] I documented the Java code using JavaDoc style.

#### Changes affecting Programming Exercises
- [x] **High priority**: I tested **all** changes and their related features with **all** corresponding user types on a test server configured with the **integrated lifecycle setup** (LocalVC and LocalCI).
- [x] I tested **all** changes and their related features with **all** corresponding user types on a test server configured with **LocalVC** and **Jenkins**.

### Motivation and Context

Artemis allows students to reset their repository in programming exercises. For individual exercises, this works fine.
However, for team exercises any of the team members can do this and there is currently no indication whatsoever who did this. The corresponding commit is attributed to Artemis and there is not even an entry in the logs about this.
This means that a malicious team member could reset the exercise just before the deadline and there is no real way to figure out who it was. This almost happened during a real team exercise.

Closes https://github.com/ls1intum/Artemis/issues/11807

### Description

- Added log output to ProgrammingExerciseParticipationResource#resetRepository.
- Modified the commit message in ProgrammingExerciseParticipationService#resetRepository to attribute the commit to the team member who reset the repository.

### Steps for Testing

Prerequisites:
- 1 Instructor
- 2 Students
- 1 Team Programming Exercise

1. Log in to Artemis as an instructor.
2. Navigate to Exercise and create a programming exercise with team participation enabled.
3. Create a team and add two test students to it
4. Log in again as a student
5. Find the newly created exercise and click "Start"
6. Then click the "Reset Repository" button
7. Check the server log, you should see a warning like: reset repo by student_login (this is easier to test locally)
reset_2

9. Clone the repository locally
10. Check the `git log` in the local repo, you should see a co-author annotation in the commit related to the reset
reset

11. Repeat steps 4–9 using the other student account
12. Now you can go back to the first test student’s local repository, run `git pull`, and check the `git log`. You’ll be able to see the reset operation performed by the second test student.

#### Exam Mode Testing

Since programming exercises in exams are single-participant only, there’s no need to test with multiple students. If you want to test it with individual participation, create an exam, then add an exercise group, and import the previously created programming exercise. Repeat steps 4–9 above to perform the test.

### Testserver States
You can manage test servers using [Helios](https://helios.aet.cit.tum.de/). Check environment statuses in the [environment list](https://helios.aet.cit.tum.de/repo/69562331/environment/list). To deploy to a test server, go to the [CI/CD](https://helios.aet.cit.tum.de/repo/69562331/ci-cd) page, find your PR or branch, and trigger the deployment.

#### Performance Review

- [x] I (as a reviewer) confirm that the server changes (in particular related to database calls) are implemented with a very good performance even for very large courses with more than 2000 students.
#### Code Review
- [x] Code Review 1
- [x] Code Review 2
#### Manual Tests
- [x] Test 1
- [x] Test 2
#### Exam Mode Test
- [x] Test 1
- [x] Test 2
#### Performance Tests
- [x] Test 1
- [x] Test 2

### Test Coverage

_Last updated: 2026-02-20 08:03:23 UTC_

## Summary by CodeRabbit

* **New Features**
  * Repository resets now include the acting user's information in the commit message and add a Co-authored-by trailer when available.
  * Reset actions record and log the identity of the user who performed the reset.
  * Performing a repository reset automatically triggers a CI build to run after the reset.

### CODE REVIEW

**Review by Hialus — DISMISSED** [review]
Code looks good 🚀

**Review by laadvo — DISMISSED** [review]
Tested on TS6, the indication who reset the repository is working as expected. I could not check the server logs in step 7, but the warning in the code seems correct.

File: src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingExerciseParticipationService.java [hunk 1]
```diff
@@ -243,9 +248,13 @@ public void resetRepository(LocalVCRepositoryUri targetUri, LocalVCRepositoryUri
                 FileUtils.copyFile(file, targetRepo.getLocalPath().resolve(file.toPath().getFileName()).toFile());
             }
         }
-
+        User coAuthor = userRepository.findOneByLogin(login).orElse(null);
+        String commitMessage = "Reset Exercise";
+        if (coAuthor != null && coAuthor.getName() != null && coAuthor.getEmail() != null) {
+            commitMessage += "\n\nCo-authored-by: " + coAuthor.getName() + " <" + coAuthor.getEmail() + ">";
+        }
```
Comment (SamuelRoettgermann) [inline src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingExerciseParticipationService.java]: I would probably change the `Co-authored-by` to smth like `Authorized by` or `Caused by`. Maybe even work with translation keys so the user gets the correct language, too (although I don't think that's necessary, as it wasn't the case before)

I would also probably loosen this a bit, so that we include what we know about the author, even if we don't know everything.

This is written free-hand in GH, so no guarantee for code correctness, but the idea should be visible:
```suggestion
if (coAuthor != null) {
    commitMessage += "\n\nResponsible person:";
    if (coAuthor.getName() == null && coAuthor.getEmail() == null) {
        commitMessage += " " + coAuthor.login;
    }
    if (coAuthor.getName() != null) {
        commitMessage += " " + coAuthor.getName();
    }
    if (coAuthor.getEmail() != null) {
        commitMessage += " <%s>".formatted(coAuthor.getEmail());
    }
}
```

File: src/main/java/de/tum/cit/aet/artemis/programming/web/ProgrammingExerciseParticipationResource.java [hunk 2]
```diff
@@ -380,7 +381,11 @@ public ResponseEntity<Void> resetRepository(@PathVariable Long participationId,
             sourceUri = exercise.getVcsTemplateRepositoryUri();
         }
 
-        programmingExerciseParticipationService.resetRepository(participation.getVcsRepositoryUri(), sourceUri);
+        String login = SecurityUtils.getCurrentUserLogin().orElse(null);
+
+        programmingExerciseParticipationService.resetRepository(participation.getVcsRepositoryUri(), sourceUri, login);
+        log.warn("Reset repo by {}", login != null ? login : "unknown");
```
Comment (SamuelRoettgermann) [inline src/main/java/de/tum/cit/aet/artemis/programming/web/ProgrammingExerciseParticipationResource.java]: Is it really necessary to do this login extraction here in the resource? Couldn't we move that to the service? That way we could also remove the `login` parameter from the `resetRepository` service function again.

The job of a resource should be input validation while actual logic should be handled by the service. In this case, getting the current user isn't really doing any input validation.

Comment (SamuelRoettgermann) [inline src/main/java/de/tum/cit/aet/artemis/programming/web/ProgrammingExerciseParticipationResource.java]: This should definitely not be of the `warn` severity. `info`, maybe even `debug` (although `info` sounds more realistic) should be the chosen severity, at least imo

**Review by SamuelRoettgermann — CHANGES_REQUESTED** [review]
some minor things I'm not fully happy with

File: src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingExerciseParticipationService.java [hunk 3]
```diff
@@ -248,13 +248,31 @@ public void resetRepository(LocalVCRepositoryUri targetUri, LocalVCRepositoryUri
...
+            if (coAuthor.getName() != null) {
+                displayIdentifier.append(" ").append(coAuthor.getName());
+            }
+            else if (coAuthor.getEmail() == null) {
+                displayIdentifier.append(" ").append(login);
+            }
+
+            if (coAuthor.getEmail() != null) {
+                displayIdentifier.append(" <").append(coAuthor.getEmail()).append(">");
+            }
```
Comment (SamuelRoettgermann) [inline src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingExerciseParticipationService.java]: Imo the logic is a bit confusing. I'd probably prefer a version with 3 separate `if`s, but this is functionally equivalent, so it's fine by me

**Review by SamuelRoettgermann — DISMISSED** [review]
Code changes look good

File: src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingExerciseParticipationService.java (hunk 1 above)
Comment (Hialus) [reply src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingExerciseParticipationService.java]: I disagree. `Co-authored-by` is a pretty standardized way of indicating who was additionally involved with a commit. This is also properly supported by every decent git client.
Any additonal information should be added on top of the standardized `Co-authored-by` instead of replacing it.

**Review by HawKhiem — DISMISSED** [review]
Other than what @SamuelRoettgermann pointed out, code looks good 👍

**Review by ShudongCai — COMMENTED** [review]
Tested on TS3, works as expected.
But I’ll wait for the final decision on the Co-authored-by wording before reapproving.

Image

File: src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingExerciseParticipationService.java (hunk 1 above)
Comment (JTNing) [reply src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingExerciseParticipationService.java]: @SamuelRoettgermann @Hialus 
I'm fine with keeping Co-authored-by for standardized authorship.
My suggestion is just to add a short note above it for destructive operations, that explicitly states the risk and the responsible person, so responsibility is clear to humans while tooling support stays intact.
```
Reset Exercise

This is a destructive operation (Repository reset).
Responsible person: NAME <EMAIL>

Co-authored-by: NAME <EMAIL>
```

Comment (JTNing) [reply src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingExerciseParticipationService.java]: Since there hasn‘t been a response for several days, I’ll proceed with using the commit message format I suggested.

Comment (SamuelRoettgermann) [reply src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingExerciseParticipationService.java]: I guess if it's standardized you can just use Co-authored instead of "Responsible person"; might actually lead to more confusion if you duplicate the same entry

**Review by ShudongCai — DISMISSED** [review]
Tested on TS4, works as expected.

File: src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingExerciseParticipationService.java [hunk 4]
```diff
@@ -248,13 +248,31 @@ public void resetRepository(LocalVCRepositoryUri targetUri, LocalVCRepositoryUri
                 FileUtils.copyFile(file, targetRepo.getLocalPath().resolve(file.toPath().getFileName()).toFile());
             }
         }
-        User coAuthor = userRepository.findOneByLogin(login).orElse(null);
+        String login = SecurityUtils.getCurrentUserLogin().orElse(null);
+        User coAuthor = login != null ? userRepository.findOneByLogin(login).orElse(null) : null;
```
Comment (SamuelRoettgermann) [inline src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingExerciseParticipationService.java:252]: Could be simplified to:
```suggestion
String login = SecurityUtils.getCurrentUserLogin().orElse(null);
User coAuthor = userRepository.findOneByLogin(login).orElse(null)
```
or straight to:
```suggestion
User coAuthor = userRepository.findOneByLogin(SecurityUtils.getCurrentUserLogin().orElse(null)).orElse(null);
```
and then later when you log the `login` (only usage of the login), you could access that value via the coAuthor variable.

**Review by SamuelRoettgermann — DISMISSED** [review]
Code looks good, only one nitpick that you don't need to fix if you don't want

**Review by laadvo — DISMISSED** [review]
Retested on ts4, the indication is still working as expected

File: src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingExerciseParticipationService.java [hunk 5]
```diff
@@ -243,9 +248,31 @@ public void resetRepository(LocalVCRepositoryUri targetUri, LocalVCRepositoryUri
...
+        if (coAuthor != null) {
+            StringBuilder displayIdentifier = new StringBuilder("\n\nCo-authored-by:");
+
+            if (coAuthor.getName() != null) {
+                displayIdentifier.append(" ").append(coAuthor.getName());
+            }
+            else if (coAuthor.getEmail() == null) {
+                displayIdentifier.append(" ").append(login);
+            }
+
```
Comment (krusche) [inline src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingExerciseParticipationService.java]: **[Medium]** Malformed Co-authored-by when name is null but email is non-null

The `else if (coAuthor.getEmail() == null)` condition means that when `name == null && email != null`, neither the name nor the login is appended. This produces:
```
Co-authored-by: <email@example.com>
```
which is a malformed Git trailer (missing the name before the email).

**Suggested fix** — always fall back to login when name is null:
```java
if (coAuthor.getName() != null) {
    displayIdentifier.append(" ").append(coAuthor.getName());
} else {
    displayIdentifier.append(" ").append(login);
}

if (coAuthor.getEmail() != null) {
    displayIdentifier.append(" <").append(coAuthor.getEmail()).append(">");
}
```
This ensures the trailer is always well-formed: `Co-authored-by: Name <email>` or `Co-authored-by: login <email>` or `Co-authored-by: login` (no email).

**Review by krusche — CHANGES_REQUESTED** [review]
**Review Summary**

Good idea to track who resets a repository — this addresses a real accountability gap in team exercises. However, there's a logic issue in the Co-authored-by formatting that produces malformed output in certain cases.

**Issues by priority:**

**Medium:**
- Malformed `Co-authored-by` trailer when `name` is null but `email` is non-null

**Low:**
- Missing server tests for the new logic

**Review by krusche — CHANGES_REQUESTED** [review]
## Review Summary

The overall intent of this PR is good -- attributing repository resets to specific users is important for accountability in team exercises. However, there are several issues that should be addressed before merging.

### Issues Found

1. **Malformed Co-authored-by trailer when email is null** (correctness): When `coAuthor.getEmail()` is null, the trailer is still appended as `Co-authored-by: Name` without the required `<email>` part. The git Co-authored-by convention requires the format `Co-authored-by: Name <email>`. A malformed trailer will not be recognized by git tooling (GitHub, GitLab, etc.) and defeats the purpose. The entire trailer should be skipped if email is unavailable, or a fallback email should be used.

2. **No tests added** (testing): This PR adds non-trivial logic (user lookup, conditional string building, null handling) but includes no unit or integration tests. The commit message construction logic with its multiple null-check branches should be tested, especially the edge cases (null email, null name, null user).

3. **StringBuilder is unnecessary overhead for simple concatenation** (code quality, minor): The `StringBuilder` with multiple appends could be replaced with a simpler `String.format("Co-authored-by: %s <%s>", name, email)` which is more readable and idiomatic for this use case.

File: src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingExerciseParticipationService.java [hunk 6]
```diff
@@ -243,9 +248,31 @@ public void resetRepository(LocalVCRepositoryUri targetUri, LocalVCRepositoryUri
...
+            if (coAuthor.getName() != null) {
+                displayIdentifier.append(" ").append(coAuthor.getName());
+            }
+            else {
+                displayIdentifier.append(" ").append(login);
+            }
+
+            if (coAuthor.getEmail() != null) {
+                displayIdentifier.append(" <").append(coAuthor.getEmail()).append(">");
+            }
```
Comment (krusche) [inline src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingExerciseParticipationService.java]: **Bug: Malformed Co-authored-by trailer when email is null**

When `coAuthor.getEmail()` is null, this code still appends the trailer resulting in something like:

```
Reset Exercise

Co-authored-by: John Doe
```

This is not a valid git trailer format. GitHub/GitLab and other tools expect `Co-authored-by: Name <email>` and will not recognize this malformed version, defeating the purpose of the attribution.

Suggestion: Only append the trailer if email is available, and simplify with `String.format`:

```java
if (coAuthor != null && coAuthor.getEmail() != null) {
    String name = coAuthor.getName() != null ? coAuthor.getName() : login;
    commitMessage += String.format("\n\nCo-authored-by: %s <%s>", name, coAuthor.getEmail());
}
```

This is simpler, always produces a well-formed trailer, and avoids the unnecessary `StringBuilder`.

**Review by krusche — CHANGES_REQUESTED** [review]

File: src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingExerciseParticipationService.java [hunk 7]
```diff
@@ -243,9 +248,19 @@ public void resetRepository(LocalVCRepositoryUri targetUri, LocalVCRepositoryUri
...
+        if (coAuthor != null && coAuthor.getEmail() != null) {
+            String name = coAuthor.getName() != null ? coAuthor.getName() : login;
+            commitMessage += String.format("\n\nCo-authored-by: %s <%s>", name, coAuthor.getEmail());
+        }
 
         gitService.stageAllChanges(targetRepo);
-        gitService.commitAndPush(targetRepo, "Reset Exercise", true, null);
+        gitService.commitAndPush(targetRepo, commitMessage, true, null);
+
+        log.info("Reset repo by {}", login != null ? login : "unknown");
```
Comment (krusche) [inline src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingExerciseParticipationService.java:263]: **[Low]** Consider using `log.warn` instead of `log.info`

Since repository resets are security-relevant actions (especially in team exercises where a malicious team member could reset before deadline, as described in the issue), using `log.warn` would make these entries easier to find in log monitoring/alerting systems.

```suggestion
        log.warn("Reset repo by {}", login != null ? login : "unknown");
```

Comment (krusche) [inline src/main/java/de/tum/cit/aet/artemis/programming/service/ProgrammingExerciseParticipationService.java:263]: **[Nitpick]** The log message could include more context

Including the repository URI or participation ID in the log message would make it easier to trace which repository was reset, especially in courses with many teams.

```suggestion
        log.info("User {} reset repository {}", login != null ? login : "unknown", targetUri);
```

**Review by krusche — APPROVED** [review]
## Review Summary

Clean and focused PR that addresses a real security concern in team programming exercises by attributing repository resets to the acting user via Git Co-authored-by trailers and server logging. The implementation is straightforward and correct.

No critical, high, or medium issues found. A couple of minor suggestions below.

**Review by ShudongCai — APPROVED** [review]
reapprove

**Review by HawKhiem — APPROVED** [review]
Reapprove

**Review by Hialus — APPROVED** [review]

**Review by Claudia-Anthropica — APPROVED** [review]
Clean, focused change that addresses a real accountability gap in team programming exercises. The null handling is solid and the Co-authored-by trailer is a nice use of standard Git conventions. Looks good.
